test/functional/shared/report_host_permissions_test.rb

```diff
@@ -0,0 +1,27 @@
+module ReportHostPermissionsTest
+  extend ActiveSupport::Concern
+  included do
+    context 'when user does not have permission to view hosts' do
+      setup { setup_user 'view', 'reports' }
+
+      test 'cannot view any reports' do
+        report = FactoryGirl.create(:report)
+        get :show, { :id => report.id }, set_session_user.merge(:user => User.current)
+        assert_response :not_found
+      end
+
+      test 'cannot view the last report' do
+        report = FactoryGirl.create(:report)
+        get :show, { :id => 'last', :host_id => report.host.id }, set_session_user.merge(:user => User.current)
```

Comment:
`:id => 'last'` is only applicable to the UI controller, the API controller has a `last` action instead.  I'm not sure sharing a test for both types of controllers is going to allow easily for these differences?